File: _REPO/APRESS/distributed-.net-programming-in-csharp/Chapter09/AsyncReceiver/ReceiverMain.cs
using System;
using System.Messaging;
using System.Runtime.Remoting.Messaging;

namespace AsyncReceiver
{
   class ReceiverMain
   {
      static void Main(string[] args)
      {
         // Open queue
         MessageQueue mq = new MessageQueue(@".\private$\MyPrivateQ");

         // Create an array of types expected in the message body
         Type[] expectedTypes = new Type[] {typeof(string), typeof(float)};

         string[] expectedTypeNames;
         expectedTypeNames = new String[] {"System.String", "System.Single"};

         // Construct formatter with expected type names
         mq.Formatter = new XmlMessageFormatter(expectedTypeNames);

         IAsyncResult ar = mq.BeginReceive(
            TimeSpan.FromSeconds(5),      // Timeout value
            mq,                           // State object, the message queue
            new AsyncCallback(OnMessageArrival) // Callback
         );

         // The MessageQueue object also provides a ReceiveCompleted event.
         // Uncomment this line to register the event handler. 
         //mq.ReceiveCompleted += new ReceiveCompletedEventHandler(OnReceiveCompleted);

         // Simulate doing other work
         while(true)
         {
            Console.WriteLine("Doing other work ...");
            System.Threading.Thread.Sleep(1000);
         }
      }

      // The callback method
      static void OnMessageArrival(IAsyncResult ar)
      {
         // Cast the state object to MessageQueue
         MessageQueue mq = (MessageQueue)ar.AsyncState;
         try
         {          
            Message msg = mq.EndReceive(ar);
            Console.WriteLine(msg.Body.ToString());
         }
         catch
         {
            Console.WriteLine("Timeout!");
         }
         finally
         {
            mq.BeginReceive( TimeSpan.FromSeconds(5), mq, 
               new AsyncCallback(OnMessageArrival));
         }
      }

      // The Receive Completed event handler
      static void OnReceiveCompleted(object sender, ReceiveCompletedEventArgs e)
      {       
         MessageQueue mq = (MessageQueue)sender;
         try
         {          
            Message msg = mq.EndReceive(e.AsyncResult);
            Console.WriteLine(msg.Body.ToString());
         }
         catch( MessageQueueException mqError)
         {
            Console.WriteLine(mqError.Message);
         }
         finally
         {
            mq.BeginReceive(TimeSpan.FromSeconds(5));
         }
      }
   }
}

File: _REPO/APRESS/distributed-.net-programming-in-vb-.net/Chapter09/AsyncReceiver/Receiver/ReceiverMain.vb
Imports System.Messaging
Imports System.Threading

Module ReceiverMain

   Sub Main()
      'Open queue
      Dim mq As New MessageQueue(".\private$\MyPrivateQ")

      'Create an array of types expected in the message body
      Dim expectedTypes() As Type 
      expectedTypes = New Type() {GetType(String), GetType(Single)}
      
      'Construct formatter with expected types
      mq.Formatter = new XmlMessageFormatter(expectedTypes)

      'Setup up the formatter (removed) ...
      
      'Param1: Timeout Value
      'Param2: State object, the message queue
      'Param3: Callback delegate
      Dim ar As IAsyncResult = mq.BeginReceive(TimeSpan.FromSeconds(5), mq, _
         new AsyncCallback(AddressOf OnMessageArrival))

      'Simulate doing other work
      Do While True
         Console.WriteLine("Doing other work on thread {0}", _
            Thread.CurrentThread.GetHashCode())
         Thread.Sleep(1000)
      Loop
   End Sub
         
   Sub OnMessageArrival(ar As IAsyncResult)

      'Cast the state object to MessageQueue
      Dim mq As MessageQueue = CType(ar.AsyncState, MessageQueue)
      Try 
               
         Dim msg As Message = mq.EndReceive(ar)
         Console.WriteLine(msg.Body.ToString())
         
      Catch
      
         Console.WriteLine("Timeout!")
         
      Finally
      
         mq.BeginReceive(TimeSpan.FromSeconds(5), mq, _
            New AsyncCallback(Addressof OnMessageArrival))
            
      End Try
   End Sub
End Module
